From bad35962b443a9b26296c6441fea460fc52e8931 Mon Sep 17 00:00:00 2001
From: Wenqiang Wei <weiwenqiang@mail.ustc.edu.cn>
Date: Thu, 10 Apr 2025 20:52:25 +0800
Subject: [PATCH 1040/1058] LATX, fix: Add option_monitor_shared_mem.

This option is used to control whether shared memory monitoring is
enabled.

Usage: export LATX_MONITOR_SHARED_MEM=1
---
 accel/tcg/cpu-exec.c                    | 2 +-
 accel/tcg/translate-all.c               | 4 +++-
 linux-user/main.c                       | 7 +++++++
 linux-user/mmap.c                       | 6 ++++--
 target/i386/latx/include/latx-options.h | 1 +
 target/i386/latx/latx-options.c         | 2 ++
 target/i386/latx/translator/translate.c | 2 +-
 7 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/accel/tcg/cpu-exec.c b/accel/tcg/cpu-exec.c
index e8203fc5e0..364b392079 100644
--- a/accel/tcg/cpu-exec.c
+++ b/accel/tcg/cpu-exec.c
@@ -261,7 +261,7 @@ cpu_tb_exec(CPUState *cpu, TranslationBlock *itb, int *tb_exit)
     env->fpu_clobber = false;
     ret = tcg_qemu_tb_exec(env, tb_ptr);
 
-    if (env->checksum_fail_tb) {
+    if (option_monitor_shared_mem && env->checksum_fail_tb) {
         TranslationBlock * tb_fail = (TranslationBlock *)env->checksum_fail_tb;
         lsassert(tb_fail->checksum && tb_fail->pc == env->eip);
         mmap_lock();
diff --git a/accel/tcg/translate-all.c b/accel/tcg/translate-all.c
index 0a5aaf4270..8920e4a002 100644
--- a/accel/tcg/translate-all.c
+++ b/accel/tcg/translate-all.c
@@ -2240,7 +2240,9 @@ TranslationBlock *tb_gen_code(CPUState *cpu,
         mprotect((void*)(pc & qemu_host_page_mask), qemu_host_page_size, PROT_READ);
     }
 
-    tb->checksum = p_flags & PAGE_MEMSHARE;
+    if (option_monitor_shared_mem) {
+        tb->checksum = p_flags & PAGE_MEMSHARE;
+    }
     gen_code_size = target_latx_host(env, tb, max_insns);
     if (unlikely(gen_code_size < 0)) {
         /*
diff --git a/linux-user/main.c b/linux-user/main.c
index ff0b40bcf9..e525886367 100644
--- a/linux-user/main.c
+++ b/linux-user/main.c
@@ -702,6 +702,11 @@ static void handle_arg_latx_real_maps(const char *arg)
     }
 }
 
+static void handle_arg_latx_monitor_shared_mem(const char *arg)
+{
+    option_monitor_shared_mem = strtol(arg, NULL, 0);
+}
+
 #ifdef CONFIG_LATX_AOT
 static void handle_arg_latx_aot(const char *arg)
 {
@@ -795,6 +800,8 @@ static const struct qemu_argument arg_table[] = {
     "",           "test memory right when memory access"},
     {"latx-real-maps",    "LATX_REAL_MAPS",     true,  handle_arg_latx_real_maps,
     "",           "enable get real self maps"},
+    {"latx-monitor-shared-mem",    "LATX_MONITOR_SHARED_MEM",     true,  handle_arg_latx_monitor_shared_mem,
+    "",           "monitor shared memory, retranslate self modifying page"},
 #ifdef CONFIG_LATX_AOT
     {"latx-aot",    "LATX_AOT",     true,  handle_arg_latx_aot,
     "",           "enable aot"},
diff --git a/linux-user/mmap.c b/linux-user/mmap.c
index fc5b77a5f2..fbad1f8c69 100644
--- a/linux-user/mmap.c
+++ b/linux-user/mmap.c
@@ -680,7 +680,9 @@ abi_long target_mmap(abi_ulong start, abi_ulong len, int target_prot,
      * be atomic with respect to an external process.
      */
     if (flags & MAP_SHARED) {
-        page_flags |= PAGE_MEMSHARE;
+        if (option_monitor_shared_mem) {
+            page_flags |= PAGE_MEMSHARE;
+        }
         CPUState *cpu = thread_cpu;
         if (!(cpu->tcg_cflags & CF_PARALLEL)) {
             cpu->tcg_cflags |= CF_PARALLEL;
@@ -1254,7 +1256,7 @@ abi_long target_mremap(abi_ulong old_addr, abi_ulong old_size,
         new_addr = h2g(host_addr);
         prot = page_get_flags(old_addr);
         page_set_flags(old_addr, old_addr + old_size, 0);
-        if ((flags & MAP_TYPE) == MAP_SHARED) {
+        if (option_monitor_shared_mem && (flags & MAP_TYPE) == MAP_SHARED) {
             prot |= PAGE_MEMSHARE;
         }
         page_set_flags(new_addr, new_addr + new_size,
diff --git a/target/i386/latx/include/latx-options.h b/target/i386/latx/include/latx-options.h
index 52a431e951..cca8719110 100644
--- a/target/i386/latx/include/latx-options.h
+++ b/target/i386/latx/include/latx-options.h
@@ -68,6 +68,7 @@ extern int option_debug_imm_reg;
 extern uint64_t imm_skip_pc;
 extern int option_mem_test;
 extern int option_real_maps;
+extern int option_monitor_shared_mem;
 
 extern unsigned long long counter_tb_exec;
 extern unsigned long long counter_tb_tr;
diff --git a/target/i386/latx/latx-options.c b/target/i386/latx/latx-options.c
index 4d59aaba17..0d51a17008 100644
--- a/target/i386/latx/latx-options.c
+++ b/target/i386/latx/latx-options.c
@@ -82,6 +82,7 @@ int option_split_tb;
 int option_anonym;
 int option_mem_test;
 int option_real_maps;
+int option_monitor_shared_mem;
 
 unsigned long long counter_tb_exec;
 unsigned long long counter_tb_tr;
@@ -131,6 +132,7 @@ void options_init(void)
     option_anonym = 0;
     option_mem_test = 0;
     option_real_maps = 0;
+    option_monitor_shared_mem = 0;
 }
 
 #define OPTIONS_IMM_REG 0
diff --git a/target/i386/latx/translator/translate.c b/target/i386/latx/translator/translate.c
index 3ca66e2ae7..aa75dc6eae 100644
--- a/target/i386/latx/translator/translate.c
+++ b/target/i386/latx/translator/translate.c
@@ -1557,7 +1557,7 @@ int tr_ir2_generate(struct TranslationBlock *tb)
     bool reduce_proepo = false;
     int tr_func_idx;
 
-    if (tb->checksum) {
+    if (option_monitor_shared_mem && tb->checksum) {
         tr_check_x86ins_change(tb);
     }
 #ifdef CONFIG_LATX_IMM_REG
-- 
2.49.0

